servers/rendering/rendering_device_driver.h

```diff
@@ -0,0 +1,663 @@
+// - We allocate as little as possible in functions expected to be quick (a counterexample is loading/saving shaders) and use alloca() whenever suitable.
+// - We try to back opaque ids with the native ones or memory addresses.
+// - Every struct has default initializers.
+// - Using VectorView to take array-like arguments. Vector<uint8_t> is an exception (an indiom for "BLOB").
+// - If a driver needs some higher-level information (the kind of info RenderingDevice keeps), it shall store a copy of what it needs.
+//   There's no backwards communication from the driver to query data from RenderingDevice.
+// ***********************************************************************************
+
+#include "core/object/object.h"
+#include "core/variant/type_info.h"
+#include "servers/display_server.h"
+#include "servers/rendering/rendering_device_commons.h"
+
+class ApiContextRD;
+
+// This may one day be used in Godot for interoperability between C arrays, Vector and LocalVector.
+// (See https://github.com/godotengine/godot-proposals/issues/5144.)
+template <class T>
+class VectorView {
+	const T *_ptr = nullptr;
+	const uint32_t _size = 0;
+
+public:
+	const T &operator[](uint32_t p_index) {
+		DEV_ASSERT(p_index < _size);
+		return _ptr[p_index];
+	}
+
+	_ALWAYS_INLINE_ const T *ptr() const { return _ptr; }
+	_ALWAYS_INLINE_ uint32_t size() const { return _size; }
+
+	VectorView() = default;
+	VectorView(const T &p_ptr) :
+			// With this one you can pass a single element very conveniently!
+			_ptr(&p_ptr),
+			_size(1) {}
+	VectorView(const T *p_ptr, uint32_t p_size) :
+			_ptr(p_ptr), _size(p_size) {}
+	VectorView(const Vector<T> &p_lv) :
+			_ptr(p_lv.ptr()), _size(p_lv.size()) {}
+	VectorView(const LocalVector<T> &p_lv) :
+			_ptr(p_lv.ptr()), _size(p_lv.size()) {}
+};
+
+// These utilities help drivers avoid allocations.
+#define ALLOCA(m_size) ((m_size != 0) ? alloca(m_size) : nullptr)
+#define ALLOCA_ARRAY(m_type, m_count) ((m_type *)ALLOCA(sizeof(m_type) * (m_count)))
+#define ALLOCA_SINGLE(m_type) ALLOCA_ARRAY(m_type, 1)
+
+// This helps forwarding certain arrays to the API with confidence.
+#define ARRAYS_COMPATIBLE(m_type_a, m_type_b) (sizeof(m_type_a) == sizeof(m_type_b) && alignof(m_type_a) == alignof(m_type_b))
+// This is used when you also need to ensure structured types are compatible field-by-field.
+// TODO: The fieldwise check is unimplemented, but still this one is useful, as a strong annotation about the needs.
+#define ARRAYS_COMPATIBLE_FIELDWISE(m_type_a, m_type_b) ARRAYS_COMPATIBLE(m_type_a, m_type_b)
+// Another utility, to make it easy to compare members of different enums, which is not fine with some compilers.
+#define ENUM_MEMBERS_EQUAL(m_a, m_b) ((int64_t)m_a == (int64_t)m_b)
+
+class RenderingDeviceDriver : public RenderingDeviceCommons {
+public:
+	struct ID {
+		uint64_t id = 0;
+		_ALWAYS_INLINE_ ID() = default;
+		_ALWAYS_INLINE_ ID(uint64_t p_id) :
+				id(p_id) {}
```

Comment:
I guess this should be changed to `size_t` to work with 32-bit builds (static assert below).

Reply:
Good point. Poor 32-bit builds, always forgotten.